File: app/src/main/java/com/vondear/tools/adapter/AdapterPopupDish.java
package com.vondear.tools.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vondear.tools.R;
import com.vondear.tools.model.ModelDish;
import com.vondear.tools.model.ModelShopCart;
import com.vondear.tools.interfaces.ShopCartInterface;

import java.util.ArrayList;

/**
 *
 * @author vondear
 * @date 16-12-23
 */
public class AdapterPopupDish extends RecyclerView.Adapter{

    private static String TAG = "PopupDishAdapter";
    private ModelShopCart mModelShopCart;
    private Context context;
    private int itemCount;
    private ArrayList<ModelDish> mModelDishList;
    private ShopCartInterface shopCartImp;

    public AdapterPopupDish(Context context, ModelShopCart modelShopCart){
        this.mModelShopCart = modelShopCart;
        this.context = context;
        this.itemCount = modelShopCart.getDishAccount();
        this.mModelDishList = new ArrayList<>();
        mModelDishList.addAll(modelShopCart.getShoppingSingleMap().keySet());
        Log.e(TAG, "PopupDishAdapter: "+this.itemCount );
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.right_dish_item1, parent, false);
        DishViewHolder viewHolder = new DishViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
        DishViewHolder dishholder = (DishViewHolder)holder;
        final ModelDish modelDish = getDishByPosition(position);
        if(modelDish !=null) {
            dishholder.right_dish_name_tv.setText(modelDish.getDishName());
            dishholder.right_dish_price_tv.setText(modelDish.getDishPrice() + "");
            int num = mModelShopCart.getShoppingSingleMap().get(modelDish);
            dishholder.right_dish_account_tv.setText(num+"");

            dishholder.right_dish_add_iv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if(mModelShopCart.addShoppingSingle(modelDish)) {
                        notifyItemChanged(position);
                        if(shopCartImp!=null)
                            shopCartImp.add(view,position);
                    }
                }
            });

            dishholder.right_dish_remove_iv.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if(mModelShopCart.subShoppingSingle(modelDish)){
                        mModelDishList.clear();
                        mModelDishList.addAll(mModelShopCart.getShoppingSingleMap().keySet());
                        itemCount = mModelShopCart.getDishAccount();
                        notifyDataSetChanged();
                        if(shopCartImp!=null)
                            shopCartImp.remove(view,position);
                    }
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return this.itemCount;
    }

    public ModelDish getDishByPosition(int position){
       return mModelDishList.get(position);
    }

    public ShopCartInterface getShopCartInterface() {
        return shopCartImp;
    }

    public void setShopCartInterface(ShopCartInterface shopCartImp) {
        this.shopCartImp = shopCartImp;
    }

    private class DishViewHolder extends RecyclerView.ViewHolder{
        private TextView right_dish_name_tv;
        private TextView right_dish_price_tv;
        private LinearLayout right_dish_layout;
        private ImageView right_dish_remove_iv;
        private ImageView right_dish_add_iv;
        private TextView right_dish_account_tv;

        public DishViewHolder(View itemView) {
            super(itemView);
            right_dish_name_tv = (TextView)itemView.findViewById(R.id.right_dish_name);
            right_dish_price_tv = (TextView)itemView.findViewById(R.id.right_dish_price);
            right_dish_layout = (LinearLayout)itemView.findViewById(R.id.right_dish_item);
            right_dish_remove_iv = (ImageView)itemView.findViewById(R.id.right_dish_remove);
            right_dish_add_iv = (ImageView)itemView.findViewById(R.id.right_dish_add);
            right_dish_account_tv = (TextView) itemView.findViewById(R.id.right_dish_account);
        }

    }
}

File: RxDemo/src/main/java/com/tamsiree/rxdemo/adapter/AdapterPopupDish.kt
package com.tamsiree.rxdemo.adapter

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tamsiree.rxdemo.R
import com.tamsiree.rxdemo.interfaces.ShopCartInterface
import com.tamsiree.rxdemo.model.ModelDish
import com.tamsiree.rxdemo.model.ModelShopCart
import java.util.*

/**
 *
 * @author tamsiree
 * @date 16-12-23
 */
class AdapterPopupDish(private val context: Context, private val mModelShopCart: ModelShopCart) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private var itemCount: Int
    private val mModelDishList: ArrayList<ModelDish>
    var shopCartInterface: ShopCartInterface? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.right_dish_item1, parent, false)
        return DishViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val dishholder = holder as DishViewHolder?
        val modelDish = getDishByPosition(position)
        dishholder!!.right_dish_name_tv.text = modelDish.dishName
        dishholder.right_dish_price_tv.text = modelDish.dishPrice.toString()
        val num = mModelShopCart.shoppingSingleMap[modelDish]!!
        dishholder.right_dish_account_tv.text = num.toString()
        dishholder.right_dish_add_iv.setOnClickListener { view ->
            if (mModelShopCart.addShoppingSingle(modelDish)) {
                notifyItemChanged(position)
                if (shopCartInterface != null) {
                    shopCartInterface!!.add(view, position)
                }
            }
        }
        dishholder.right_dish_remove_iv.setOnClickListener { view ->
            if (mModelShopCart.subShoppingSingle(modelDish)) {
                mModelDishList.clear()
                mModelDishList.addAll(mModelShopCart.shoppingSingleMap.keys)
                itemCount = mModelShopCart.dishAccount
                notifyDataSetChanged()
                if (shopCartInterface != null) {
                    shopCartInterface!!.remove(view, position)
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return itemCount
    }

    fun getDishByPosition(position: Int): ModelDish {
        return mModelDishList[position]
    }

    private inner class DishViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val right_dish_name_tv: TextView
        val right_dish_price_tv: TextView
        private val right_dish_layout: LinearLayout
        val right_dish_remove_iv: ImageView
        val right_dish_add_iv: ImageView
        val right_dish_account_tv: TextView

        init {
            right_dish_name_tv = itemView.findViewById<View>(R.id.right_dish_name) as TextView
            right_dish_price_tv = itemView.findViewById<View>(R.id.right_dish_price) as TextView
            right_dish_layout = itemView.findViewById<View>(R.id.right_dish_item) as LinearLayout
            right_dish_remove_iv = itemView.findViewById<View>(R.id.right_dish_remove) as ImageView
            right_dish_add_iv = itemView.findViewById<View>(R.id.right_dish_add) as ImageView
            right_dish_account_tv = itemView.findViewById<View>(R.id.right_dish_account) as TextView
        }
    }

    companion object {
        private const val TAG = "PopupDishAdapter"
    }

    init {
        itemCount = mModelShopCart.dishAccount
        mModelDishList = ArrayList()
        mModelDishList.addAll(mModelShopCart.shoppingSingleMap.keys)
        Log.e(TAG, "PopupDishAdapter: " + itemCount)
    }
}